test/cpp/qps/driver.cc

```diff
@@ -504,16 +508,19 @@ std::unique_ptr<ScenarioResult> RunScenario(
       GPR_ASSERT(false);
     }
   }
-  gpr_log(GPR_INFO, "Finishing servers");
-  for (size_t i = 0; i < num_servers; i++) {
-    auto server = &servers[i];
-    if (!server->stream->Write(server_mark)) {
-      gpr_log(GPR_ERROR, "Couldn't write mark to server %zu", i);
-      GPR_ASSERT(false);
-    }
-    if (!server->stream->WritesDone()) {
-      gpr_log(GPR_ERROR, "Failed WritesDone for server %zu", i);
-      GPR_ASSERT(false);
+
+  if (!client_finish_first) {
+    gpr_log(GPR_INFO, "Finishing servers");
```

Comment:
There's so much common content here between the two cases; do you want to move this to a side function?

Reply:
Sure I'm going to refactor this large function into many small functions.